File: src/credentials_rotators/npm/lambda_functions/npm_utils.py
import json
import logging

import pyotp

import requests

logger = logging.getLogger()
logger.setLevel(logging.INFO)

def generate_otp(username, otp_seed):
    """Generate a time based OTP using the OTP_SEED for the npm user
    It is used in 2 Factor Authentication for the user account
    Args:
        username (string): The username of the npm user
        otp_seed (string): The seed generated during 2 factor auth setup for the user
    """
    totp = pyotp.parse_uri(f'otpauth://totp/{username}?secret={otp_seed}&issuer=npm')
    otp = totp.now()
    return otp


def update_login_password(username, otp_seed, current_password, new_password):
    """Update the login password for the npm user
    Args:
        username (string): The username of the npm user
        otp_seed (string): The seed generated during 2 factor auth setup for the user
        current_password (string): The current active login password for the npm user
        new_password (string): The new login password to be set for the npm user
    Raises:
        HttpError: If the password update fails
    """
    otp = generate_otp(username, otp_seed)

    headers = {
        'content-type': 'application/json',
        'npm-otp': otp
    }

    url = 'https://registry.npmjs.org/-/npm/v1/user'
    data_dict = {'password': {'old': current_password, 'new': new_password}}
    data = json.dumps(data_dict)

    response = requests.post(url, headers=headers, data=data, auth=(username, current_password))
    response.raise_for_status()

def create_access_token(username, otp_seed, password):
    """Create an access token for a NPM user
    Args:
        username (string): The username of the npm user
        otp_seed (string): The seed generated during 2 factor auth setup for the user
        password (string): The current active login password for the npm user
    Raises:
        HttpError: If the access token creation fails
    """
    otp = generate_otp(username, otp_seed)

    headers = {
        'content-type': 'application/json',
        'npm-otp': otp
    }

    url = 'https://registry.npmjs.org/-/npm/v1/tokens'
    data_dict = {'password': password}
    data = json.dumps(data_dict)

    response = requests.post(url, headers=headers, data=data, auth=(username, password))
    response.raise_for_status()
    return json.loads(response.content)['token']


def get_user_info_using_password(username, otp_seed, password):
    """Fetch the npm user profile information after authorization using password and OTP
    Args:
        username (string): The username of the npm user
        otp_seed (string): The seed generated during 2 factor auth setup for the user
        password (string): The current active login password for the npm user
    Raises:
        HttpError: If the user profile information cannot be fetched
    """
    otp = generate_otp(username, otp_seed)

    headers = {
        'content-type': 'application/json',
        'npm-otp': otp
    }

    url = 'https://registry.npmjs.org/-/npm/v1/user'
    response = requests.get(url, headers=headers, auth=(username, password))
    response.raise_for_status()

def get_user_info_using_access_token(username, otp_seed, access_token):
    """Fetch the npm user profile information after authorization using access token and OTP
    Args:
        username (string): The username of the npm user
        otp_seed (string): The seed generated during 2 factor auth setup for the user
        access_token (string): The active access token for the npm user
    Raises:
        HttpError: If the user profile information cannot be fetched
    """
    otp = generate_otp(username, otp_seed)

    headers = {
        'Authorization': f'Bearer {access_token}',
        'content-type': 'application/json',
        'npm-otp': otp
    }

    url = 'https://registry.npmjs.org/-/npm/v1/user'
    response = requests.get(url, headers=headers)
    response.raise_for_status()

def delete_access_token(username, otp_seed, password, access_token):
    """Delete given access token for a NPM user
    Args:
        username (string): The username of the npm user
        otp_seed (string): The seed generated during 2 factor auth setup for the user
        password (string): The current active login password for the npm user
        access_token (string): The access_token to be deleted
    Raises:
        HttpError: If the access token creation fails
    """
    otp = generate_otp(username, otp_seed)

    headers = {
        'content-type': 'application/json',
        'npm-otp': otp
    }

    url = f'https://registry.npmjs.org/-/npm/v1/tokens/token/{access_token}'
    data_dict = {'password': password}
    data = json.dumps(data_dict)

    response = requests.delete(url, headers=headers, data=data, auth=(username, password))
    response.raise_for_status()

File: src/credentials_rotators/npm-token-rotation/src/lambda/utils/npm-utils.ts
import TOTP from "totp-generator";
import axios from "axios";
import { URL } from "url";
import assert from 'assert';

export const generateOtp = (otpUrl: string) => {
  let seed;
  if (otpUrl.startsWith("otpauth")) {
    const url = new URL(otpUrl);
    seed = url.searchParams.get("secret");
    assert(seed, "OTP URL is missing secret parameter");
  } else {
    seed = otpUrl;
  }
  return TOTP(seed, { digits: 6 });
};

export const createAccessToken = async (
  username: string,
  password: string,
  otpSeed: string
): Promise<string> => {
  try {
    const otp = generateOtp(otpSeed);
    const result = await axios.post(
      "https://registry.npmjs.org/-/npm/v1/tokens",
      {
        password,
      },
      {
        headers: {
          "content-type": "application/json",
          "npm-otp": otp,
        },
        auth: {
          username,
          password,
        },
      }
    );
    return result.data["token"];
  } catch (e) {
    throw new Error(`Error occurred when creating token: ${e.message}`);
  }
};

export const deleteAccessToken = async (
  username: string,
  password: string,
  otpSeed: string,
  token: string
): Promise<boolean> => {
  try {
    const otp = generateOtp(otpSeed);
    const result = await axios.delete(
      `https://registry.npmjs.org/-/npm/v1/tokens/token/${token}`,
      {
        headers: {
          "content-type": "application/json",
          "npm-otp": otp,
        },
        auth: {
          username,
          password,
        },
      }
    );
    return result.status >= 200 && result.status < 300;
  } catch (e) {
    throw new Error(`Error occurred when deleting token: ${e.message}`);
  }
};

export const validateAccessToken = async (
  username: string,
  accessToken: string
): Promise<boolean> => {
  try {
    const result = await axios.get("https://registry.npmjs.org/-/npm/v1/user", {
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "content-type": "application/json",
      },
    });
    return result.status >= 200 && result.status < 300;
  } catch (e) {
    throw new Error(`Error occurred when verifying token: ${e.message}`);
  }
};
